src/swap/useSwapQuote.ts

```diff
@@ -1,27 +1,238 @@
+          e
+        )
+        continue
+      }
+      throw e
+    }
+
+    const swapTx = {
+      ...baseSwapTx,
+      maxFeePerGas,
+      maxPriorityFeePerGas,
+      feeCurrency: feeCurrencyAddress,
+      // We assume the provided gas value is with the native fee currency
+      // If it's not, we add the static padding
+      gas: !feeCurrency.isNative ? baseSwapTx.gas + BigInt(STATIC_GAS_PADDING) : baseSwapTx.gas,
+    }
+
+    const maxGasCost = getMaxGasCost([approveTx, swapTx])
+    const maxGasCostInDecimal = new BigNumber(maxGasCost.toString()).shiftedBy(
+      -feeCurrency.decimals
+    )
+    maxGasCosts.push({ feeCurrency, maxGasCostInDecimal })
+    if (maxGasCostInDecimal.isGreaterThan(feeCurrency.balance)) {
+      // Not enough balance to pay for gas, try next fee currency
+      continue
+    }
+
+    const fromAmount = new BigNumber(approveTx.amountToApprove.toString()).shiftedBy(
+      -fromToken.decimals
+    )
+    if (
+      fromToken.tokenId === feeCurrency.tokenId &&
+      fromAmount.plus(maxGasCostInDecimal).isGreaterThan(fromToken.balance)
+    ) {
+      // Not enough balance to pay for gas, try next fee currency
+      continue
+    }
+
+    // This is the one we can use
+    return {
+      type: 'possible',
+      approveTransaction: approveTx,
+      swapTransaction: swapTx,
+    } satisfies PreparedTransactionsPossible
+  }
+
+  // So far not enough balance to pay for gas
+  // let's see if we can decrease the swap from amount
+  const result = maxGasCosts.find(({ feeCurrency }) => feeCurrency.tokenId === fromToken.tokenId)
+  if (!result || result.maxGasCostInDecimal.isGreaterThan(result.feeCurrency.balance)) {
+    // Can't decrease the swap from amount
+    return {
+      type: 'not-enough-balance-for-gas',
+      feeCurrencies,
+    } satisfies PreparedTransactionsNotEnoughBalanceForGas
+  }
+
+  // We can decrease the swap from amount to pay for gas,
+  // We'll ask the user if they want to proceed
+  const adjustedMaxGasCost = result.maxGasCostInDecimal.times(
+    DECREASED_SWAP_AMOUNT_GAS_COST_MULTIPLIER
+  )
+  const maxFromAmount = fromToken.balance.minus(adjustedMaxGasCost)
+  const maxToAmount = maxFromAmount.times(price)
+
+  return {
+    type: 'need-decrease-swap-amount-for-gas',
+    maxGasCost: adjustedMaxGasCost,
+    feeCurrency: result.feeCurrency,
+    decreasedSwapAmount: updatedField === Field.FROM ? maxFromAmount : maxToAmount,
+  } satisfies PreparedTransactionsNeedDecreaseSwapAmountForGas
 }
 
 const useSwapQuote = () => {
   const walletAddress = useSelector(walletAddressSelector)
   const useGuaranteedPrice = useSelector(guaranteedSwapPriceEnabledSelector)
-  const [exchangeRate, setExchangeRate] = useState<ExchangeRate | null>(null)
+  const [exchangeRate, setExchangeRate] = useState<QuoteResult | null>(null)
+
+  // TODO use the networkId from the fromToken
```

Comment:
same questions as the other todo. In general, I worry that todo's are easily forgotten about

Reply:
Yes fair enough, we have work queued up for multichain support. This will be changed in a follow up PR.